### PR TITLE
kubelet should report observed generation in status of bound pods

### PR DESCRIPTION
**Pull request by mikedanese**
Visibility into a flake: #21367 and a generally useful thing to have.

### CODE REVIEW

File: pkg/kubelet/status/manager.go [hunk 1]
```diff
@@ -247,6 +247,9 @@ func (m *manager) updateStatusInternal(pod *api.Pod, status api.PodStatus, force
 		oldStatus = pod.Status
 	}
 
+	// update observed generation
+	status.ObservedGeneration = pod.ObjectMeta.Generation
```
Comment (mikedanese) [inline pkg/kubelet/status/manager.go]: I'd like to report observed generation for every sync of a PodSpec. This seems like a good place to do it.

**Comment by mikedanese** [comment]
cc @bgrant0607 @kubernetes/goog-node

**Comment by yujuhong** [comment]
Commented on the issue: https://github.com/kubernetes/kubernetes/issues/21367#issuecomment-192526218

**Comment by dchen1107** [comment]
Issue # please? what kind of problem you are trying to solve here?

**Comment by dchen1107** [comment]
Ahh, just saw @yujuhong refers this to #21367

**Comment by bgrant0607** [comment]
The API and registry bits LGTM, other than the comment I just asked you to add.

**Comment by yujuhong** [comment]
> The API and registry bits LGTM, other than the comment I just asked you to add.

Given the implementation today, this doesn't help deflaking the test. Quoting myself in https://github.com/kubernetes/kubernetes/issues/21367#issuecomment-192526218: "I don't think this will work as you expected. kubelet only sends a status update for a pod if there is a status change. You won't be able to get the generation to confirm until that happens (and it may never happen until the pod is deleted) .How about accessing the /pods endpoint on kubelet to get the list of pods it sees?"

**Comment by mikedanese** [comment]
> kubelet only sends a status update for a pod if there is a status change

We are modifying the status by incrementing the pod.status.observedGeneration so the change should trigger a reconcile:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go#L447

we should trigger an update of status when the spec is updated either here:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go#L347

or here:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1727

what am I missing?

**Comment by yujuhong** [comment]
Discussed with @bgrant0607 @dchen1107 @mikedanese offline: It's valuable to have the `ObservedGeneration`, but the consensus is that we want the `ObservedGeneration` to be what kubelet used to sync the pod. 
In short, (1) it won't help the test if we implement it this way, and (2) this needs a bit of change in kubelet and is not needed for v1.2.

**Comment by bgrant0607** [comment]
In ReplicationController, ReplicaSet, and Deployment controllers, observedGeneration is updated when the new spec has been taken into account and the status accurately reflects that change. The change doesn't need to be fully reified. 

For example, someone might increase the number of ReplicaSet replicas by 1000. That requires creating lots of pods, in multiple bursts. The controller will create some, then update status. It's acceptable to update observedGeneration at that time, even though not all the pods have been created and not all creations have even been observed.

Maybe a more relevant example: The pod template of a Deployment is changed. The Deployment controller sees this and sets updatedReplicas to 0, since no replicas of the latest pod template exist as yet. It's acceptable to update observedGeneration at that point, since that informs a client that when updatedReplicas matches replicas, all the pods will be using the updated template, not the previous one.

In this case, I'd expect the status to reflect the change to the pod. For instance, if the image were changed and a new image needed to be pulled, the container would be reported as Waiting before observedGeneration was updated.

**Comment by mikedanese** [comment]
Or in the case of graceful termination, SIGTERM has been sent event though graceful termination may not have completed yet.

**Pull request closed without merging**